feat: add media type examples override decorator

IgorKarpiuk opened the pull request:
## What/Why/How?

Add media type examples override decorator

## Reference
closes: https://github.com/Redocly/redocly-cli/issues/440
## Testing

## Screenshots (optional)

## Check yourself

- [x] Code is linted
- [ ] Tested with redoc/reference-docs/workflows
- [x] All new/updated code is covered with tests

## Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

tatomyr commented on packages/core/src/decorators/common/media-type-examples-override.ts:
```diff
@@ -0,0 +1,41 @@
+import { readAndParseFileSync } from '../../utils';
+
+export const MediaTypeExamplesOverride: Oas3Decorator = ({ operationIds }) => {
+  let operationId: string | undefined;
```
Do you think it should be here? Let's try to remove it.

tatomyr commented on packages/core/src/decorators/common/media-type-examples-override.ts:
```diff
@@ -0,0 +1,41 @@
+    Operation: {
+      leave(operation: Oas3Operation) {
+        operationId = operation.operationId;
+        const properties = operationIds[operationId!];
```
I'd put some kind of guard instead of '!'.

tatomyr commented on packages/core/src/decorators/common/media-type-examples-override.ts:
```diff
@@ -0,0 +1,41 @@
+        operationId = operation.operationId;
+        const properties = operationIds[operationId!];
+
+        if (properties.responses && operation.responses) {
```
Could you use a uniform approach both here and [here](https://github.com/Redocly/redocly-cli/pull/897/files#diff-150f23c15f2f75bc7e9556b690b68cd5c77fa61ce8801d6fd346ea86775c26feR20)?

tatomyr commented on packages/core/src/decorators/common/media-type-examples-override.ts:
```diff
@@ -0,0 +1,41 @@
+
+            Object.values(operation.responses[responseCode].content!).forEach((mimeType) => {
+              if (mimeType.examples) {
+                mimeType.examples = readAndParseFileSync(properties.responses[responseCode]);
```
I'd suggest figuring out what type of example field is used, something like this:

```ts
const exampleField = mimeType.examples ? 'examples' : 'example';
mimeType[exampleField] = readAndParseFileSync(properties.responses[responseCode]);
```

tatomyr commented on packages/core/src/decorators/common/media-type-examples-override.ts:
```diff
@@ -0,0 +1,41 @@
+              continue;
+            }
+
+            Object.values(operation.responses[responseCode].content!).forEach((mimeType) => {
```
Could you use either `forEach` or `for ... of` here and [here](https://github.com/Redocly/redocly-cli/pull/897/files#diff-150f23c15f2f75bc7e9556b690b68cd5c77fa61ce8801d6fd346ea86775c26feR33)?

tatomyr commented on packages/core/src/decorators/common/media-type-examples-override.ts:
```diff
@@ -0,0 +1,39 @@
+
+        if (properties.request && operation.requestBody) {
+          Object.keys(properties.request).forEach((mimeType) => {
+            (operation.requestBody as Oas3RequestBody).content[mimeType].example =
```
Would it work when we're referencing say, `components`?

IgorKarpiuk replied:
Yes

adamaltman commented on packages/core/src/decorators/__tests__/medi-type-examples-override.test.ts:
```diff
@@ -0,0 +1,432 @@
+import { makeConfig, parseYamlToDocument, yamlSerializer } from '../../../__tests__/utils';
```
The file name has a typo in "media".

adamaltman commented on packages/core/src/decorators/__tests__/medi-type-examples-override.test.ts:
```diff
@@ -0,0 +1,432 @@
+                description: json
+                content:
+                  application/json:
+                    example:
```
This seems like it is a mistake. I think this means to be `examples`?

IgorKarpiuk replied:
Yes, thanks

adamaltman reviewed (changes requested):
This should always place target examples into the `examples` map and never into `example`. The examples map accepts a value in a specific structure (Example object)[https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#media-type-object] whereas the `example` field expects the raw example itself.

I didn't see the most common use case tested, which is no examples defined and then the override adds an example to the examples map.

IgorKarpiuk commented:
> This should always place target examples into the `examples` map and never into `example`. The examples map accepts a value in a specific structure (Example object)[https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#media-type-object] whereas the `example` field expects the raw example itself.
> 
> I didn't see the most common use case tested, which is no examples defined and then the override adds an example to the examples map.

@adamaltman I changed `example` to `examples` now it will be always place target examples into the `examples` field and i added new tests

adamaltman commented on packages/core/src/decorators/__tests__/media-type-examples-override.test.ts:
```diff
@@ -0,0 +1,555 @@
+    `);
+  });
+
+  it('should not override the examples', async () => {
```
This is the wrong behavior. It should create the examples.

adamaltman replied:
How is the config currently?

IgorKarpiuk replied:
@adamaltman 

Now it looks like this: 
```yaml
decorators:
  media-type-examples-override:
    operationIds:
       PostPets:
          request: 
              application/json: ./pet-examples.yaml
          responses: 
            '200': ./pet-examples.yaml
            '400': ./pet-errors-examples.yaml
```

adamaltman replied:
Yes, you're right.

IgorKarpiuk replied:
Changed ✅

adamaltman reviewed (changes requested):
The behavior should be different. When you want to add an example and there isn't an example there yet, it should allow you to do that. The test I highlighted seems to do the opposite behavior.

adamaltman reviewed (approved):
You should also add docs. You can look at the other decorator docs and copy the page structure and change the details.

tatomyr commented on packages/core/src/utils.ts:
```diff
@@ -153,6 +153,11 @@ export function readFileAsStringSync(filePath: string) {
   return fs.readFileSync(filePath, 'utf-8');
 }
 
+export function readAndParseFileSync<T>(filePath: string) {
```
Is it only for yaml and json files? If so, it worth renaming this functions to be explicit.

tatomyr commented on packages/core/src/decorators/common/media-type-examples-override.ts:
```diff
@@ -0,0 +1,79 @@
+import { ResolveFn, UserContext } from '../../walk';
+
+export const MediaTypeExamplesOverride: Oas3Decorator = ({ operationIds }) => {
+  function checkAndResolveRef<T>(node: any, resolver: ResolveFn): T | undefined {
+    if (!isRef(node)) {
+      return node;
+    }
+
+    const resolved = resolver<T>(node);
+    return resolved.error ? undefined : JSON.parse(JSON.stringify(resolved.node));
+  }
```
Please move this function out of the decorator.

tatomyr commented:
Left a few minor comments.
